### PR TITLE
Abort the MPI network upon unhandled exception.

### PR DESCRIPTION
**Pull request by gbalduzz**
I realized that both with the old master and the current version, if a single processor throws an unhandled exception (i.e. by reading an input file or by running out of memory), and the stack unwinds until MPI_Finalize is called, only that process terminates, while the rest hangs on the first MPI communication. 
This PR shuts down all the processes unless a different behaviour is specified (used for selectively killing processors upon driver failure).
If you have a better implementation in mind, please say so.

### CODE REVIEW

**Comment by uhaehner** [comment]
test this please

File: src/parallel/mpi_concurrency/mpi_concurrency.cpp [hunk 1]
```diff
@@ -26,6 +26,7 @@ MPIConcurrency::MPIConcurrency(int argc, char** argv) : MPIInitializer(argc, arg
       error_string = "Process " + std::to_string(MPIProcessorGrouping::get_world_id()) +
                      "could not execute a CUDA kernel.";
 
+    MPIInitializer::continueOnException();
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply src/parallel/mpi_concurrency/mpi_concurrency.cpp]: What I want to achieve is to select when a failure should halt only the current processes, and when to abort the entire computation. When I throw this exception, I know that the processor is outside the MPIGrouping communicator, and the rest of the processor will be carry on doing more measurements. Also if I go past the MPI_Finalize call, whatever I do later, will not affect the rest of the processors no matter what. So the only sensible thing to do with a processor that fails to create a Concurrency object is to unwind the stack and exit.
 If any other exception happens later, there is yet no mechanism to distribute the computation to the rest of the processors, and I want a failure of any process inside MPIGrouping, to propagate to the rest.

[the remark this answers is not in the document]
Comment (biddisco) [reply src/parallel/mpi_concurrency/mpi_concurrency.cpp]: I guess it depends on how likely it is that an exception is going to cause a big code to hang. I'd say, the main DCA program/s that are going to be run at scale where a hanging application would be a trauma should be modified to handle exceptions, but small tests etc can be ignored for now as they will be picked up by timeouts and can be modified as needed.

**Comment by gbalduzz** [comment]
I changed approach in 3d19098f4abb2ef92e43c5779c235743386d72b6 by placing the main of distributed applications, concurrency aside, inside a try-catch bracket.  If you have a better solution please say so.

**Comment by biddisco** [comment]
Actually. I found a problem. Don't merge this yet. Compilation with MPI turned off is failing.

**Comment by uhaehner** [comment]
retest this please

**Pull request merged**